### ISSUE TITLE
Incorrect implementation of Not and And matchers

### ISSUE THREAD

**Issue by abacabadabacaba** [tavianator/irregex#1]
The implementation of `Not` and `And` matchers is incorrect, as demonstrated by the following tests:

```rust
assert_matches(r"(!a)*", &["aaa"], &[]);
assert_matches(r"((a|b)&(ab?))*", &[], &["ab"]);
```

In fact, your code claims to match a pattern of length $m$ against a string of length $n$ in $O(mn)$ time, but this complexity is probably not possible to achieve if a pattern can contain `Not` and/or `And`.

**Comment by tavianator** [comment]
Yep this was pointed out to me last night but I had to go to sleep.  I'll fix the post soon.
